Update "Build an App" examples: use secure cluster

rmloveland opened the pull request:
Addresses #631

Summary of changes:

- Update code samples for the following languages/ORMs to connect to secure local cluster using certs:
    - Java JDBC
    - Java Hibernate
    - Ruby pg
    - Ruby ActiveRecord
    - Python psycopg2
    - Python SQLAlchemy
    - Node.js pg

- Update cluster startup instructions for the above languages to use secure practices

- Create a new versioned include directory for 2.1 docs, which holds all secure code and cluster setup instructions

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/3271)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/b56e0a5041ff1af9ee52d7126a761186c92cbc96/

BramGruneir commented:
Sure.  I'll take a look shortly.  Reviewable seems to be down.  :(

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a few comments.  

---

Reviewed 24 of 24 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[_includes/v2.1/app/common-steps-secure.md, line 9 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSeaD8-3SWsUYNEF9B:-LFSeaD9-kDF7nuYpyXB:bq0x929) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/common-steps-secure.md#L9)):*
> ```Markdown
> First, [create the certificates](create-security-certificates.html), including for the user `maxroach` who will be connecting to the database:
> 
> {% include copy-clipboard.html %}
> ```

Can you add a comment as to what each of these commands are doing?  I know it's in the other doc page, but people are lazy.

---

*[_includes/v2.1/app/common-steps-secure.md, line 30 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSenjL05fTkmYz4y5A:-LFSenjL05fTkmYz4y5B:b7xzwtk) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/common-steps-secure.md#L30)):*
> ```Markdown
> 
> Next, [start the local cluster](secure-a-cluster.html):
> ```

same here, a few comments to quickly describe what the command is doing and how each line differs.

---

*[_includes/v2.1/app/common-steps-secure.md, line 54 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSf1J3EtuPm9a94tgu:-LFSf1J3EtuPm9a94tgv:b-zgldwy) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/common-steps-secure.md#L54)):*
> ```Markdown
> 
> In a new terminal, as the `root` user, use the [`cockroach user`](create-and-manage-users.html) command to create a new user, `maxroach`.
> ```

So is it just me, or is it weird that you can create a cert for a user that doesn't exist yet?

---

*[_includes/v2.1/app/txn-sample.js, line 151 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSfItQ9u2P5MpCGE_R:-LFSfItQ9u2P5MpCGE_S:b-ho1ub0) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/txn-sample.js#L151)):*
> ```JavaScript
>             finish();
>         });
> });
> ```

missing blank line at the end of the file

---

*[_includes/v2.1/app/hibernate-basic-sample/build.gradle, line 15 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSf_I270tbJ8EaVkxx:-LFSf_I270tbJ8EaVkxy:b-hhxql3) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/hibernate-basic-sample/build.gradle#L15)):*
> ```Gradle
> dependencies {
>     compile 'org.hibernate:hibernate-core:5.2.4.Final'
>     compile 'org.postgresql:postgresql:9.4.1208'
> ```

Should we not move this up to 9.5?

---

*[_includes/v2.1/app/hibernate-basic-sample/hibernate-basic-sample.tgz, line 0 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSffs19Cuwktk9CUqu:-LFSffs19Cuwktk9CUqv:b-duu7fl) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/hibernate-basic-sample/hibernate-basic-sample.tgz#L0)):*
Why are you checking in a zipped file here?

---

*[v2.1/build-a-java-app-with-cockroachdb.md, line 38 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSg21J8jqgSS8g8SQu:-LFSg21K6JKX3JmWuc5Q:b-awnhxh) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/v2.1/build-a-java-app-with-cockroachdb.md#L38)):*
> ```Markdown
> 
> {% include copy-clipboard.html %}
> ~~~ shell
> ```

please add a comment as to what this does and why it is necessary.

---

*[v2.1/build-a-java-app-with-cockroachdb-hibernate.md, line 62 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSgCRZ8vdMW-ybYaX6:-LFSgCRZ8vdMW-ybYaX7:b4pah40) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/v2.1/build-a-java-app-with-cockroachdb-hibernate.md#L62)):*
> ```Markdown
> ## Step 5. Convert the key file for use with Java
> 
> In the directory where you store your certificates (`/tmp/certs` in this example), convert maxroach's user key file into a format that will work with Java like so:
> ```

Please add a few more details here on what this doing.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3271#-:-LFShSxeDl8-1jGRyexN:bwnnd54)*
<!-- Sent from Reviewable.io -->

rmloveland commented:
Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[_includes/v2.1/app/common-steps-secure.md, line 9 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSeaD8-3SWsUYNEF9B:-LFXtj2G-hKP9zPp86Re:b-l0t6ei) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/common-steps-secure.md#L9)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Can you add a comment as to what each of these commands are doing?  I know it's in the other doc page, but people are lazy.
</blockquote></details>

Nice catch! Fixed.

---

*[_includes/v2.1/app/common-steps-secure.md, line 30 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSenjL05fTkmYz4y5A:-LFXuXPo2DkWFrW6GsDW:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/common-steps-secure.md#L30)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

same here, a few comments to quickly describe what the command is doing and how each line differs.
</blockquote></details>

Done.

---

*[_includes/v2.1/app/common-steps-secure.md, line 54 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSf1J3EtuPm9a94tgu:-LFXucia2UaqS_P95dth:b-ftsc7k) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/common-steps-secure.md#L54)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

So is it just me, or is it weird that you can create a cert for a user that doesn't exist yet?
</blockquote></details>

Interesting, I didn't even consider that.  I suppose from one POV it's nice to be able to script the cert creation before getting the DB up, but ... yeah, huh.  Perhaps there are reasons this could be bad. (IDK - I am not a security expert or even a novice really.)

I will see if I can find out more about this in #cli or some other channel.  My guess is that the cert will be a useless credential if the user is never created in the DB?  but again, IDK :-) 

Update: apparently you can connect with the cert as that user even if the `CREATE USER` has not been run - they also don't show up in `SHOW USERS`.  I don't know if this is a bug but it's the behavior Raphael described to me (see #cli for details as I may have misunderstood somewhat)

---

*[_includes/v2.1/app/txn-sample.js, line 151 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSfItQ9u2P5MpCGE_R:-LFYU4Q54kmcfjw0MHm3:bj3eoea) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/txn-sample.js#L151)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

missing blank line at the end of the file
</blockquote></details>

Fixed, thanks.

---

*[_includes/v2.1/app/hibernate-basic-sample/build.gradle, line 15 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSf_I270tbJ8EaVkxx:-LFYVzYa8bhxMEv04KUy:b-xx2us4) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/hibernate-basic-sample/build.gradle#L15)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Should we not move this up to 9.5?
</blockquote></details>

Good call. Upgrading to very latest JDBC postgres driver required adding the SQL dialect to the Hibernate config, so I added that too (in both text file and the - ugh, sorry git! - gzip).

---

*[_includes/v2.1/app/hibernate-basic-sample/hibernate-basic-sample.tgz, line  at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSffs19Cuwktk9CUqu:-LFXwsfvCY2lCiGup5Cq:b-m5f5f9) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/_includes/v2.1/app/hibernate-basic-sample/hibernate-basic-sample.tgz#L)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Why are you checking in a zipped file here?
</blockquote></details>

I agree this is very strange.  We are already providing a zipped file for the previous "insecure" versions of this page using this method, so I thought it would be the least invasive/smallest change to do the same with the "secure" page until we are able to work on moving code samples out to repos in future as we have discussed.  FTR I don't like it though (and I know git doesn't like it either).

---

*[v2.1/build-a-java-app-with-cockroachdb.md, line 38 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSg21J8jqgSS8g8SQu:-LFXxdky0jn1C4dZo00d:b-tkdx7a) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/v2.1/build-a-java-app-with-cockroachdb.md#L38)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

please add a comment as to what this does and why it is necessary.
</blockquote></details>

Ach! Fixed.  Thank you.  This of all commands is the source of the most heartache and needs explanation. :-)

---

*[v2.1/build-a-java-app-with-cockroachdb-hibernate.md, line 62 at r1](https://reviewable.io/reviews/cockroachdb/docs/3271#-LFSgCRZ8vdMW-ybYaX6:-LFYST8g9IpM9dBsSLtP:b-9xxhg4) ([raw file](https://github.com/cockroachdb/docs/blob/b56e0a5041ff1af9ee52d7126a761186c92cbc96/v2.1/build-a-java-app-with-cockroachdb-hibernate.md#L62)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please add a few more details here on what this doing.
</blockquote></details>

Fixed, using same explanation as the other Java page.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/docs/3271)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f81e1ecadf74584e87db30586ec86c6b6e0ce973/

Amruta-Ranade commented on _includes/v2.1/app/common-steps-secure.md:
```diff
@@ -0,0 +1,84 @@
+1. Create a CA (certificate authority) that will be used to sign all of your other certificates.
+
+    {% include copy-clipboard.html %}
+    ~~~ sh
```
Normally we ask users to create the certs directory and then create the certificates within that directory. I am curious as to why did we change the directory to /tmp/certs/ and if we should change it elsewhere as well.

Amruta-Ranade replied:
Also, copy-pasting and running the command gives an error because the directory doesn't exist. I know it's a simple error that the user can fix, but just FYI

Amruta-Ranade replied:
`--certs-dir=/tmp/certs--certs-dir=/tmp/certs` should be `--certs-dir=/tmp/certs`

Amruta-Ranade commented on _includes/v2.1/app/common-steps-secure.md:
```diff
@@ -0,0 +1,84 @@
+    $ cockroach cert create-ca --certs-dir=/tmp/certs--certs-dir=/tmp/certs --ca-key=/tmp/certs/ca.key
+    ~~~
+
+2. Create a certificate and key for each node in the cluster.  Since we are running a local cluster, we will cheat a bit by re-using the same certificate and key for each node.
```
Well, we aren't really cheating. We are just being technically correct :) I think we can simply say: "Since we are running a local cluster, we can reuse..."

Amruta-Ranade commented on _includes/v2.1/app/common-steps-secure.md:
```diff
@@ -0,0 +1,84 @@
+
+5. Start the nodes that will make up the local cluster by running the following commands.
+
+    {% include copy-clipboard.html %}
```
Nit: The commands need to be run on separate terminals.

Amruta-Ranade commented:
@rmloveland This PR's awesome 🎉 
I have tested the python docs (sqlalchemy and psycopg2). Have yet to test the other code samples, but wanted to get the feedback for the python code samples in so you wouldn't have to wait longer. Sorry for the delay!

Amruta-Ranade commented on v2.1/build-a-ruby-app-with-cockroachdb-activerecord.md:
```diff
@@ -33,22 +33,20 @@ To install ActiveRecord as well as the [pg driver](https://rubygems.org/gems/pg)
-{{site.data.alerts.callout_success}}
-The exact command above will vary depending on the desired version of ActiveRecord. Specifically, version 4.2.x of ActiveRecord requires version 0.1.x of the adapter; version 5.1.x of ActiveRecord requires version 0.2.x of the adapter.
-{{site.data.alerts.end}}
+{{site.data.alerts.callout_success}}The exact command above will vary depending on the desired version of ActiveRecord. Specifically, version 4.2.x of ActiveRecord requires version 0.1.x of the adapter; version 5.1.x of ActiveRecord requires version 0.2.x of the adapter.{{site.data.alerts.end}}
```
How do I get the 0.2.x version of the adapter? 
(Here's the error I get: 
in `raise_if_conflicts': Unable to activate activerecord-cockroachdb-adapter-0.2.3, because pg-1.0.0 conflicts with pg (< 0.22, >= 0.20) (Gem::ConflictError))

rmloveland replied:
I got this error too (but I did not get it several weeks ago, huh).  Strangely, when you do `gem install pg` it installs both 0.2 and 1.0 -- at least on my machine.

Was able to fix this and get it working by changing the order of the imports in the file `activerecord-basic-sample.rb` to list the `pg` module last.  🤷‍♂️  Yay computers!

Amruta-Ranade commented on v2.1/build-a-ruby-app-with-cockroachdb.md:
```diff
@@ -39,11 +39,11 @@ Now that you have a database and a user, you'll run code to create a table and i
 
 First, use the following code to connect as the `maxroach` user and execute some basic SQL statements, creating a table, inserting rows, and reading and printing the rows.
```
This sentence confuses me. When I parse the sentences: "First, use the following code..." followed by "Download the file...", I first try to find the code to connect using the maxroach user (and so on), and only then am I supposed to download the file.

I suggest we use the phrasing in the ActiveRecord section: "The following code uses...." to describe what the code in the file does.

Amruta-Ranade commented on v2.1/build-a-ruby-app-with-cockroachdb.md:
```diff
@@ -65,15 +65,13 @@ Initial balances:
+{% include v2.1/app/txn-sample.rb %}
 ~~~
 
 Then run the code:
```
Getting this error:

/usr/local/lib/ruby/gems/2.4.0/gems/pg-1.0.0/lib/pg.rb:56:in `initialize': fe_sendauth: no password supplied (PG::ConnectionBad)
	from /usr/local/lib/ruby/gems/2.4.0/gems/pg-1.0.0/lib/pg.rb:56:in `new'
	from /usr/local/lib/ruby/gems/2.4.0/gems/pg-1.0.0/lib/pg.rb:56:in `connect'
	from txn-sample.rb:36:in `<main>'

rmloveland commented on _includes/v2.1/app/common-steps-secure.md:
```diff
@@ -0,0 +1,84 @@
+    $ cockroach cert create-ca --certs-dir=/tmp/certs--certs-dir=/tmp/certs --ca-key=/tmp/certs/ca.key
+    ~~~
+
+2. Create a certificate and key for each node in the cluster.  Since we are running a local cluster, we will cheat a bit by re-using the same certificate and key for each node.
```
Fixed.

rmloveland commented on _includes/v2.1/app/common-steps-secure.md:
```diff
@@ -0,0 +1,84 @@
+
+    {% include copy-clipboard.html %}
+    ~~~ sh
+    $ cockroach cert create-ca --certs-dir=/tmp/certs--certs-dir=/tmp/certs --ca-key=/tmp/certs/ca.key
```
> I am curious as to why we create the certs in a /tmp/certs directory?

Good point, updated to do `mkdir certs` and then use that.  This should fix the error when copy-pasting the code.  Thanks for catching that.

> --certs-dir=/tmp/certs--certs-dir=/tmp/certs needs to be --certs-dir=/tmp/certs

Also fixed, thanks!

rmloveland commented:
> Need help testing the java docs.

@Amruta-Ranade would you mind sharing the commands you're running and the errors that result?

(Aside: This might be a sign we need to update [Build a Java App](http://cockroachlabs.com/docs/dev/build-a-java-app-with-cockroachdb.html) to make it more explicit/easier to follow)

rmloveland commented:
@jseldess I will follow up with Amruta to help with testing the Java stuff (as discussed above), but I think this is ready for a 2nd review.

As soon as this one is ready to merge, I'll have another PR coming quickly behind with more secure examples.  At that point we may want to discuss whether it's more valuable to "securify" the rest of the examples (not sure how many people are writing CRDB-using apps directly in C/C++/Rust) or move over to ORM txn/concurrency issues since that seems to be affecting more users.

jseldess commented:
Rich and I discussed offline that these pages should offer both secure and insecure options. We'll try to remove the cluster start step from these tutorials and instead use the "Before you begin" section to route people to the correct page and have them identify what secure mode they're using so we can provide the correct guidance and code:

```
## Before you begin

1. [Install CockroachDB](install-cockroachdb.html).
2. Start up a [secure](secure-a-cluster.html) or [insecure](start-a-local-cluster.html) local cluster.
3. Identify which security mode you're using:
    [ Secure ] [ Insecure ]
```

where `[ Secure ] [ Insecure ]` uses the div approach to show/hide content like on this page: https://raw.githubusercontent.com/cockroachdb/docs/master/_includes/v2.1/orchestration/start-kubernetes.md.

rmloveland commented:
> Rich and I discussed offline that these pages should offer both secure and insecure options

@jseldess this is ready for another round of reviews when you get a chance.  I just pushed changes to all of the pages that add a "secure/insecure" toggle to the instructions, as we discussed.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e3910a06992653800e10b08f719fada679423bb2/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/be9327928d77f0f3eb4c86e3bc195867ddba75ad/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/7c46395fdff6f98021b0e74bd6cc50431d589dc5/
